### DISCUSSION TITLE
ASGI and RSGI headers format

### DISCUSSION THREAD

**Discussion by cirospaciari** [emmett-framework/granian#15] [Q&A] [answered]
Trying to Test ASGI + Falcon with this code:

```python
import falcon.asgi
from granian import Granian


class Home:
    async def on_get(self, req, resp):
        resp.status = falcon.HTTP_200  # This is the default status
        resp.content_type = falcon.MEDIA_TEXT  # Default is JSON, so override
        resp.text = "Hello, World!"
    async def on_post(self, req, resp):
        raw_data = await req.stream.read()
        print("data", raw_data)
        resp.status = falcon.HTTP_200  # This is the default status
        resp.content_type = falcon.MEDIA_TEXT  # Default is JSON, so override
        resp.text = raw_data
        
        


app = falcon.asgi.App()

home = Home()
app.add_route("/", home)

if __name__ == "__main__":
    Granian(
        f"falcon-asgi:app",
        address="0.0.0.0",
        port=4001,
        workers= 1, #multiprocessing.cpu_count(),
        backlog=2048,
        interface="asgi",
        websockets=False
    ).serve()
```

I got the error
```bash
  File "/usr/local/lib/python3.10/dist-packages/falcon/asgi/request.py", line 393, in __init__
    for header_name, header_value in scope['headers']:
ValueError: too many values to unpack (expected 2)
```

The headers for ASGI are incompatible, in uvicorn they are an array of tuples, and in the ASGI docs:

For HTTP:

>"headers (Iterable[[byte string, byte string]]) – An iterable of [name, value] two-item iterables, where name is the header name, and value is the header value. Order of header values must be preserved from the original HTTP request; order of header names is not important. Duplicates are possible and must be preserved in the message as received. Header names should be lowercased, but it is not required; servers should preserve header case on a best-effort basis. Pseudo headers (present in HTTP/2 and HTTP/3) must be removed; if :authority is present its value must be added to the start of the iterable with host as the header name or replace any existing host header already present."

https://asgi.readthedocs.io/en/latest/specs/www.html#http-connection-scope

For WebSockets:

>"headers (Iterable[[byte string, byte string]]) – An iterable of [name, value] two-item iterables, where name is the header name and value is the header value. Order should be preserved from the original HTTP request; duplicates are possible and must be preserved in the message as received. Header names should be lowercased, but it is not required; servers should preserve header case on a best-effort basis. Pseudo headers (present in HTTP/2 and HTTP/3) must be removed; if :authority is present its value must be added to the start of the iterable with host as the header name or replace any existing host header already present."

https://asgi.readthedocs.io/en/latest/specs/www.html#websocket-connection-scope

**Comment by gi0baro** [comment, marked as answer]
@cirospaciari this is a bug now tracked in #16

**Reply by cirospaciari** [reply]
Thanks for the Quick fix! I will also try to implement RSGI for some projects, really cool interface
